man/man1/tabgen.1
.TH TABGEN 1NEMO "10 April 2026"

.SH "NAME"
tabgen \- Create an arbitrarely large ASCII table with (random) values

.SH "SYNOPSIS"
\fBtabgen\fP [parameter=value]

.SH "DESCRIPTION"
Create a table with (random) numbers.
Either uniformly ditributed numbers between 0 and 1,
from a gaussian (normal) distribution with mean 0 and sigma 1, 
a constant value, or a linearly increasing value (0, 1, 2, ...). See \fBmode=\fP below.
.PP
The number or rows and columns can be arbitrarely large, but not all \fItable(5NEMO)\fP
programs in NEMO will be able to deal with large columns or rows. Usually rows
can be managed with the \fBnmax=\fP keyword, but columns is hidden deeper and not under
user control.
.PP
Using \fBmode < 0\fP
numbers are generated but not output, in effect allowing a benchmark.
.PP
 This program has no real use other than benchmarking and testing the limits
of other program being fed very large tables. See EXAMPLES below
for a counter example.

.SH "PARAMETERS"
.so man1/parameters
.TP 20
\fBout=\fP
Output ascii table [???]     
.TP
\fBnr=\fP
Number of rows [10]    
.TP
\fBnc=\fP
Number of columns [5]    
.TP
\fBmode=\fP
The distribution of numbers: 1=uniform   2=normal   3=constant  4=linear
< 0 : no output, but the
absolute value determines which distribution of numbers  [1]
.TP 
\fBseed=\fP
Random seed [123]
.TP
\fBfmt=\fP
The \fIprintf(3)\fP style format specifier. [%g]
.TP
\fBsep=c|s|v|t\fP
Separator between columns. c=comma  s=space  v=vertical bar  t=tab.
If none of these are given, the string is taken literal
[Default: space]
.TP
\fBaddrow=t|f\fP
add a row counter, 1 being the first. [f]   

.SH "PERFORMANCE"
On a typical 2020 machine it takes 0.23 seconds to create and output 1M random numbers,
however most of the time is the output function. Using mode < 0 you can test the speed
of just producing random numbers. About 95% of the time is spent in \fIprintf(3)\fP,
as can be seen with this benchmark:
.EX

    /usr/bin/time tabgen . 10000 10000 -1
    1.38user 0.29system 0:01.68elapsed 99%CPU
    
    /usr/bin/time tabgen . 10000 10000 1
    28.60user 0.14system 0:28.77elapsed 99%CPU
    
.EE
The raw production of random numbers is thus about 72 Mrups (same speed as numpy.random.normal),
where the Linpack benchmark claims this machine runs at 50 GFlops.
.PP
Generating a sequence of numbers is relatively slow in tabgen.
To generate 10M numbers is slow compared
ot the standard unix tool \fIseq(1)\fP, however
the new \fBaddrow=\fP parameter makes it about 3x faster.
.EX

    /usr/bin/time tabgen - 10000000 1 4 fmt=%-10.f > tab2
    2.07user 0.10system 0:02.18elapsed 99%CPU

    /usr/bin/time seq 10000000 > tab2
    0.05user 0.04system 0:00.10elapsed 100%CPU (

but

    /usr/bin/time tabgen - 10000000 0 addrow=t > tab2
    0.65user 0.07system 0:00.73elapsed 99%CPU
    
.EE
.PP
The \fInemoinp(1NEMO)\fP program can also generate a simple table, but
the maximum number of elements in an implied list is limited (100,000 currently).
This is where \fBtabgen\fP might be useful:
.EX

     nemoinp 1:200000 > tab1
     ### Fatal error [nemoinp]: Too many items in list, use a bigger nmax=100000
     
     tabgen - 200000 1 mode=4 > tab1
     wc -l tab1
     200000 tab1

.EE


.SH "EXAMPLES"
An example showing off Unix pipes: generate a table, sorting the first column,
and smoothing and plotting the second column divided by the first column, which winds up
showing an interesting \fI"spectrum"\fP:
.EX

tabgen - 100 2 |\\
     sort -n |\\
     tabmath - - %1/%2 |\\
     tabsmooth - 3 tcol=1 |\\
     tabplot - line=1,1

.EE

An example of turning a matrix in row based single column:
.EX

tabgen - nr=20 nc=5 mode=4 | tabcols - colsep=n

.EE
Should list "100" as the last element.

.SH "SEE ALSO"
seq(1), tabhead(1NEMO), tabmath(1NEMO), xrandom(1NEMO), nemoinp(1NEMO), xrandom(3NEMO), printf(3)

.SH "AUTHOR"
Peter Teuben

.SH "HISTORY"
.nf
.ta +1.5i +5.5i
12-Jul-2020	V0.1 Created	PJT
14-jul-2020	V0.3 mode= implemented	PJT
25-jul-2020	V0.4 changed meaning of mode=	PJT
29-apr-2022	V0.6 cleanup and tableV2 things 	PJT
5-oct-2022	V0.7 added sep=	PJT
14-feb-2024	V0.8 added addrow=	PJT
.fi


src/kernel/tab/tabgen.c
/*
 *  TABGEN:     generate a table with (random) numbers, purely for benchmarking
 */

#include <nemo.h>

string defv[] = {
    "out=???\n     Output table",
    "nr=10\n       Number of rows",
    "nc=5\n        Number of columns",
    "mode=1\n      Mode (1=uniform, 2=normal 3=constant 4=linear, 5=linear)",
    "seed=123\n    Random seed",
    "fmt=%g\n      Format statement for output",
    "sep=s\n       column separator (s=space t=tab c=comma v=vertical bar)",
    "addrow=f\n    Add row number (1=first) as first column?",
    "VERSION=last",
    "header=None\n     Add a dummy header in a given style [ecsv, ipac]",
    NULL,
};

string usage="Create a table with (random) numbers";


typedef real (*my_real_proc)(real,real);

local real constant(real c1, real c2)
{
  return c2;
}

local real linear(real c1, real c2)
{
  static real count = 0;
  count = count + 1;
  return count;
}

string add_header(int nc)
{
  char *result = malloc(101);
  result[0] = '\0';
  int first = 1;
  int arr_len = 0;

  strcat(result, "| ");
  for (int i = 0; i < nc; i++) {
        char col_num[100];
        if (!first) {
            strcat(result, " | ");
        }
        sprintf(col_num, "col%d", (i+1));
        strcat(result, col_num);
        first = 0;
        arr_len++;
  }

  strcat(result, " |");
  strcat(result, "\n");
  strcat(result, "| double");

  for (int i = 1; i < arr_len; i++) {
        strcat(result, " | double");
  }

  strcat(result, " |");

  return result;

}

void nemo_main()
{
  stream ostr = stropen(getparam("out"),"w");
  int i, nr = getiparam("nr");
  int j, nc = getiparam("nc");
  int mode = getiparam("mode");
  int amode = ABS(mode);
  int seed = init_xrandom(getparam("seed"));
  string fmt = getparam("fmt");
  my_real_proc my_random = NULL;
  string seps = getparam("sep");
  bool Qrow = getbparam("addrow");
  char sep[8];
  string header = getparam("header");

  fprintf(ostr, "%s\n", add_header(nc));

  if (seps[0] == 'c') strcpy(sep,",");
  else if (seps[0] == 's') strcpy(sep," ");
  else if (seps[0] == 't') strcpy(sep,"\t");
  else if (seps[0] == 'v') strcpy(sep,"|");
  else strcpy(sep,seps);

  dprintf(1,"seed=%d\n",seed);

  if (amode == 1) {
    dprintf(1,"Uniform values between 0 and 1\n");
    my_random = xrandom;
  } else if (amode == 2) {
    dprintf(1,"Uniform values between 0 and 1\n");
    my_random = grandom;
  } else if (amode == 3) {
    dprintf(1,"Uniform values between 0 and 1\n");
    my_random = constant;
  } else if (amode == 4) {
    dprintf(1,"Uniform values between 0 and 1\n");
    my_random = linear;
  } else
    error("Illegal mode=%d",mode);

  if (mode < 0) {
    // only produce random numbers, testing production rate
    real *x = (real *) allocate(nc*nr*sizeof(real));
    int k=0;
    for (i=0; i<nr; i++)
      for (j=0; j<nc; j++)
        x[k++] = my_random(0.0,1.0);
  } else
    // output to file as well
    for (i=0; i<nr; i++) {
      if (Qrow) fprintf(ostr,"%d ", i+1);
      for (j=0; j<nc; j++) {
        if (j>0) fprintf(ostr, "%s", sep);
        fprintf(ostr,fmt, my_random(0.0, 1.0));
      }
      fprintf(ostr,"\n");
    }

}